cw/util/trace_library.py

```diff
@@ -0,0 +1,194 @@
+from typing import Optional
+
+import numpy as np
+import sqlalchemy as db
+import sqlalchemy_utils as db_utils
+from sqlalchemy.orm import sessionmaker
+
+
+@dataclass
+class Metadata:
+    """ Metadata for capture.
+
+    Stores information about the capture of the traces.
+
+    """
+    config: str
+    datetime: datetime.datetime
+    bitstream: str
+    binary: str
+    offset: Optional[int] = 0
+    sample_rate: Optional[int] = 0
+    scope_gain: Optional[float] = 0
+    trigger_level: Optional[float] = 0
+    time_div: Optional[float] = 0
+
+
+@dataclass
+class Trace:
+    """ Trace.
+
+    Defines the trace format.
+
+    """
+    wave: bytearray
+    plaintext: bytearray
+    ciphertext: bytearray
+    key: bytearray
+    x_pos: Optional[int] = 0
+    y_pos: Optional[int] = 0
+    trace_id: Optional[int] = 0
+
+
+class TraceLibrary:
+    """ Class for writing/reading traces to a database.
+
+    Either a new database is created or traces are appended to the existing
+    one. Traces are first written into memory and the flushed into the
+    database after reaching a trace memory threshold.
+    """
+    def __init__(self, db_name, trace_threshold):
+        self.engine = db.create_engine("sqlite:///" + db_name + ".db")
+        db_utils.create_database(self.engine.url)
+        self.Session = sessionmaker(self.engine)
+        self.session = self.Session()
+        self.metadata = db.MetaData()
+        self.TracesDB = db.Table(
+            "traces",
+            self.metadata,
+            db.Column("trace_id", db.Integer, primary_key=True,
+                      autoincrement=True),
+            db.Column("wave", db.LargeBinary),
+            db.Column("plaintext", db.LargeBinary),
+            db.Column("ciphertext", db.LargeBinary),
+            db.Column("key", db.LargeBinary),
+            db.Column("x_pos", db.Integer),
+            db.Column("y_pos", db.Integer),
+        )
+        self.MetadataDB = db.Table(
+            "metadata",
+            self.metadata,
+            db.Column("config", db.String),
+            db.Column("datetime", db.DateTime),
+            db.Column("bitstream", db.String),
+            db.Column("binary", db.String),
+            db.Column("offset", db.Integer),
+            db.Column("sample_rate", db.Integer),
+            db.Column("scope_gain", db.Float),
+            db.Column("trigger_level", db.Float),
+            db.Column("time_div", db.Float)
+        )
+        self.metadata.create_all(self.engine)
+        self.trace_mem = []
+        self.trace_mem_ctr = 0
+        self.trace_mem_thr = trace_threshold
+        self.settingsDict = {}
+
+    def commit_traces(self):
+        """ Writes traces from memory into database.
+        """
+        if len(self.trace_mem):
+            query = db.insert(self.TracesDB)
+            traces = []
+            for trace in self.trace_mem:
+                tr = asdict(trace)
+                del tr["trace_id"]
+                traces.append(tr)
```

Comment:
I wonder if it would be simpler to do this conversion inside `write_trace`. If I understand correctly, we only use `trace_mem` as an input for this function, so converting the traces to dictionaries and dropping the unused ids as the traces get appended might make things slightly simpler.

Reply:
The idea is to store traces in memory (=`trace_mem`) to make things faster. When writing traces to the storage (database), we do not really care about speed. Therefore, the conversion is currently done in the `commit_traces `function. However, in reality, I do not think that there is a big performance impact. But I keep it as it is for now.

Reply:
Ahah! I hadn't twigged that we might not actually do the conversion every time. I'm convinced :-)